lib/src/vs/oracle.rs

```diff
@@ -0,0 +1,167 @@
+	//
+	// Second, it can be not monotonic if the database is restarted within a second,
+	// because the in-memory counter resets to 0 on restart and the physical time uses the glanularity
+	// of seconds.
+	//
+	// Do also note that the produced versionstamp's monotonicity needs to be guaranteed
+	// by the caller of the oracle too.
+	// For example, if it is going to be used as the commit versionstamp of a transaction,
+	// the caller needs to ensure that the commit versionstamp is always increasing.
+	// Otherwise, the transaction might be committed with a versionstamp that is smaller than
+	// the versionstamp of a previous transaction.
+	// This is because the versionstamp oracle does not know the context of the versionstamp
+	// and cannot guarantee the monotonicity of the versionstamp by itself.
+	// Implementation-wise, this can imply that the database needs to ensure that the only one
+	// transaction is executing and committing at a time.
+	SysTimeCounter(SysTimeCounter),
+	// EpochCounter versionstamp oracle provides a vector clock that uses the "epoch" as the first logical time
+	// and the in-memory counter for the second logical time.
+	//
+	// The epoch is supposed to be persisted in the underlying KVS and increased by one on each database restart.
+	// The in-memory counter resets on each database restart and increased by one on each now() call.
+	//
+	// EpochCounter is designed to be used instead of the SysTimeCounter when the runtime environment
+	// does not provide a monotonic system clock, and the database is running in a single-node mode.
+	EpochCounter(EpochCounter),
+}
+
+impl Oracle {
+	pub fn now(&mut self) -> Versionstamp {
+		match self {
+			Oracle::SysTimeCounter(sys) => sys.now(),
+			Oracle::EpochCounter(epoch) => epoch.now(),
+		}
+	}
+}
+
+pub struct SysTimeCounter {
+	// The first element is the saved physical time of the last versionstamp.
+	// The second element is the in-memory counter that resets every second.
+	state: Mutex<(u64, u16)>,
+
+	stale: (u64, u16),
+}
+
+impl SysTimeCounter {
+	pub fn now(&mut self) -> Versionstamp {
+		// Increment the counter and get the current number as the logical time
+		// only if the current physical time is the same as the last physical time.
+		// Otherwise, reset the counter to 0 and get the current number as the logical time.
+		// This is to ensure that the logical time is always increasing.
+		let state = self.state.lock();
+		if let Ok(mut state) = state {
+			let (last_physical_time, counter) = *state;
+			let current_physical_time = secs_since_unix_epoch();
+			let current_logical_time = if last_physical_time == current_physical_time {
+				counter
+			} else {
+				state.1 = 0;
+				0
+			};
+			state.0 = current_physical_time;
+			state.1 += 1;
+			self.stale = (current_physical_time, current_logical_time);
+			u64_u16_to_versionstamp(current_physical_time, current_logical_time)
+		} else {
+			u64_u16_to_versionstamp(self.stale.0, self.stale.1)
```

Comment:
Does this mean in case when a mutex lock fails, we return back older version of timestamp? Wouldn't that be inconsistent as it would be returning an already existing timestamp which could have been used?

Reply:
@arriqaaq Thanks! You are correct! And, yes, they can be inconsistent.

But do note that even though this oracle provided no stale verionstamps, you still need to separately ensure that the transactions are committed with consistent order and unique motonic versionstamps.

I.e. you need to separately ensure that the commits are made in the order of unique versionstamps and I presume that's not strictly the responsibility of this package.

That said, I presume we have two other options; Do you prefer one of those or have another idea?

- Return `Err(PleaseRetryTheFunCall)` on L12 so that the caller of `now()` can retry calling it until succeed (with or without aborting the whole transaction, depends on how you impl it). The resulting versionstamps are guaranteed to be monotonic and unique.
- Make this `async` and wait until the lock succeeds. The resulting versionstamps are guaranteed to be monotonic and unique.

Reply:
Thank you for the explanation, I understand the challenges here. I think both approaches are nice, I strongly don't have a preference because as you've said, it is better to pass on the responsibility to the caller on how to handle it.